numpy/fft/_pocketfft_umath.c

```diff
@@ -0,0 +1,354 @@
+    npy_intp si = steps[0], sf = steps[1], so = steps[2];
+    npy_intp nin = dimensions[1], nout = dimensions[2];
+    npy_intp isi = steps[3], iso = steps[4];
+    npy_intp npts = nout;
+    char *buffer = NULL;
+    int no_mem = 1;
+
+    rfft_plan plan = make_rfft_plan(npts);
+    if (plan == NULL) {
+        goto fail;
+    }
+    if (iso != sizeof(npy_double)) {
+        buffer = malloc(npts * sizeof(npy_double));
+        if (buffer == NULL) {
+            goto fail;
+        }
+    }
+
+    for (npy_intp i = 0; i < n_outer; i++, ip += si, fp += sf, op += so) {
+        double fct = *(double *)fp;
+        char *op_buf = buffer == NULL? op : buffer;
+        double *op_double = (double *)op_buf;
+        /* Copy complex input to buffer in FFTpack order */
+        op_double[0] = ((double *)ip)[0];
+        if (nin > 1) {
+            copy_data(ip + isi, isi, nin - 2,
+                      (char *)&op_double[1], sizeof(npy_cdouble), npts / 2 - 1,
+                      sizeof(npy_cdouble));
+            npy_intp ncopied = (npts / 2 - 1  < nin - 2 ? npts / 2 - 1 : nin - 2);
+            double *last = (double *)(ip + (ncopied + 1) * isi);
+            op_double[ncopied * 2 + 1] = last[0];
+            if (npts % 2 == 1) {  /* For odd n, we still imag real of the last point */
+                op_double[ncopied * 2 + 2] = last[1];
+            }
+        }
+        if ((no_mem = rfft_backward(plan, op_double, fct)) != 0) {
+            break;
+        }
+        if (op_buf != op) {
+            copy_data(op_buf, sizeof(npy_double), npts, op, iso, npts, sizeof(npy_double));
+        }
+    }
+  fail:
+    if (buffer != NULL) {
+        free(buffer);
+    }
+    if (plan != NULL) {
+        destroy_rfft_plan(plan);
+    }
+    if (no_mem) {
+        /* TODO: Requires use of new ufunc API to indicate error return */
+        NPY_ALLOW_C_API_DEF
+        NPY_ALLOW_C_API;
+        PyErr_NoMemory();
+        NPY_DISABLE_C_API;
+    }
+    return;
+}
+
+
+static PyUFuncGenericFunction fft_functions[] = { fft_loop };
+static char fft_types[] = { NPY_CDOUBLE, NPY_DOUBLE, NPY_CDOUBLE};
+static void *fft_data[] = { &cfft_forward };
+static void *ifft_data[] = { &cfft_backward };
+
+static PyUFuncGenericFunction rfft_n_even_functions[] = { rfft_n_even_loop };
+static PyUFuncGenericFunction rfft_n_odd_functions[] = { rfft_n_odd_loop };
+static char rfft_types[] = { NPY_DOUBLE, NPY_DOUBLE, NPY_CDOUBLE};
+static void *rfft_data[] = { (void *)NULL };
+
+static PyUFuncGenericFunction irfft_functions[] = { irfft_loop };
+static char irfft_types[] = { NPY_CDOUBLE, NPY_DOUBLE, NPY_DOUBLE};
+static void *irfft_data[] = { (void *)NULL };
+
+static int
+add_gufuncs(PyObject *dictionary) {
+    PyObject *f;
+
+    f = PyUFunc_FromFuncAndDataAndSignature(
+        fft_functions, fft_data, fft_types, 1, 2, 1, PyUFunc_None,
+        "fft", "complex forward FFT\n", 0, "(n),()->(m)");
+    /* yes, this should not happen, but helps debugging. */
```

Comment:
Checking for error is the clean thing to do.  So I would say: delete the comment, and add the check to all ufunc creations.

`SetItemString` can (theoretically) also fail.

Reply:
OK, added the check to all functions (but not to `SetItemString` - if that fails on module import, there are bigger problems...)